# Week 02 – Glassbox Models (part 2)

In this notebook we'll be applying different glassbox models to the same dataset and compare their performance in terms of accuracy and f1-score and their interpretability.

#### Step 1: Navigate to this week's directory 
```
cd <path_to_week_2_material>
```

#### Step 2: Create and Activate a virtual environment (Python 3.9)
**MacOS**
```
python3.11 -m venv part2_venv
source part2_venv/bin/activate
```

**Windows (cmd)**
```
python3.11 -m venv part2_venv
part2_venv\Scripts\activate
```

#### Step 3: Install required packages
First, install `ipykernel` to integrate your virtual environment with Jupyter.
```
pip install ipykernel
python -m ipykernel install --user --name=part2_venv
```

Next, install all necessary packages.
```
pip install ruleopt
```

**import packages**

In [1]:
import warnings
warnings.filterwarnings("ignore")
import numpy as np
import pandas as pd
from sklearn.metrics import accuracy_score, f1_score, confusion_matrix, roc_auc_score
import os
import sys
# import matplotlib.pyplot as plt

# Load the dataset

We are using the [Titanic dataset](https://www.kaggle.com/c/titanic/overview), which holds data about passangers of the Titanic and whether they survived or not. Passengers are described by 7 features. The response variable is binary (0 – died; 1 - survived).

Make sure to save the dataset in the parent directory or adjust the file path below.

We are directly loading the pre-processed data sets. For some algorithms, we require the data to be in binary form. Hence, we have two versions of X: `X_train` and `X_test`, with continous features and one-hot encoded categorical features, and `X_train_bin` and `X_test_bin`, where all features have been one-hot encoded. For this, continuous features were first transformed into categories. Check the code in `./01_intro/titanic_data_prep.ipynb` for more details on pre-processing.

In [2]:
# X_train 
X_train = pd.read_csv('../datasets/titanic/encoded_titanic_X_train.csv')
X_train_bin = pd.read_csv('../datasets/titanic/bin_titanic_X_train.csv')

# X_test
X_test = pd.read_csv('../datasets/titanic/encoded_titanic_X_test.csv')
X_test_bin = pd.read_csv('../datasets/titanic/bin_titanic_X_test.csv')

# y_train and y_test
y_train = pd.read_csv('../datasets/titanic/titanic_y_train.csv')
y_test = pd.read_csv('../datasets/titanic/titanic_y_test.csv')

# take a look at the data
X_train.head()

Age  SibSp  Parch     Fare  Sex_1  Embarked_1  Embarked_2  Pclass_1  \
0  29.699118      0      0   7.6292      0           1           0         0   
1  29.699118      0      0   8.0500      1           0           1         0   
2  29.699118      0      0   7.7500      0           1           0         0   
3  51.000000      1      0  77.9583      0           0           1         0   
4  21.000000      0      0   7.7333      1           1           0         0   

   Pclass_2  
0         1  
1         1  
2         1  
3         0  
4         1

## About the dataset

- `Age` – age of a person in years (int)
- `SibSp` – the number of siblings or spouse of a person **onboard** (int)
- `Parch` – the number of parents or children of a person **onbard** (int)
- `Fare` – ticket price (float)
- `Sex` – sex of a person (categorical/binary)
- `Embarked` – location where the traveler mounted from. There are three possible values — Southampton, Cherbourg, and Queenstown (categorical)
- `Pclass` – passenger division into class 1, 2, and 3 (categorical)
- `Survived` – whether person survived the sinking of the ship (binary). Less than 40% survived. This is the **outcome** to predict. 


From the original dataset and from preprocessing the data, we know the following about the **categorical features**:

- `Sex` has two values `['female','male']`, which were encoded `[0,1]`, respectively. Then, after applying one-hot encoding, we have `Sex_1` which indicates `male` if 1, `female` otherwise.
- `Embarked` has three values `['C', 'Q', 'S']`, which were encoded `[0,1,2]`, respectively. Hence,
    - `Embarked_1 = 1` indicates `Q` 
    - `Embarked_2 = 1` indicates `S`
    - `Embarked_1 = 0` and `Embarked_2 = 0` indicated `C`
- `Pclass` has three values `[1,2,3]`, which were encoded `[0,1,2]`, respectively. Hence, after encoding, we have:
    - `Pclass_1 = 1` indicates `2`
    - `Pclass_2 = 1` indicates `3`
    - `Pclass_1 = 0` and `Pclass_2 = 0` indicates `1`

In [3]:
d = {
    'target': 'Survived',
    'numerical':['Age', 'SibSp', 'Parch', 'Fare'],
    'categorical':['Sex', 'Embarked', 'Pclass']
}

# save feature names
feature_names = X_train.columns
target_names = list(y_train[d['target']].unique())

# create a dataframe to save y values in 
y_results = pd.DataFrame()
y_results['y_test'] = y_test

---
---

# 5. Decision Rules

## 5.2 Rule Generation (RUG)

We propose this algorithm in our paper which you can access here: https://arxiv.org/abs/2104.10751 

In [4]:
# sys.path.insert(1, os.path.dirname(os.getcwd())+'/RuleDiscovery')
from ruleopt import RUGClassifier
from ruleopt.rule_cost import Length, Gini
from ruleopt.explainer import Explainer
# import ruxg

In [5]:
# solver = ORToolsSolver()
rule_cost = Length()

# Initialize the RUGClassifier with specific parameters
rug = RUGClassifier(
    random_state=100,
    max_rmp_calls=8,
    rule_cost=rule_cost,
    max_depth=3,
    threshold=0.05)

In [6]:
# rug = RUGClassifier(max_depth=3, rule_length_cost=False,
#                     solver='gurobi', random_state=0, max_RMP_calls = 8, threshold = 0.05)
rug.fit(X_train, y_train[d['target']])
y_results['rug_pred'] = rug.predict(np.array(X_test))

## inspect performance

In [7]:
# Confusion matrix
cm = pd.crosstab(y_results['y_test'], y_results['rug_pred'])
print ("Confusion matrix : \n", cm)

print('\nAccuracy  = %.4f' % accuracy_score(y_results['y_test'], y_results['rug_pred']))
print('F1 score  = %.4f' % f1_score(y_results['y_test'], y_results['rug_pred']))

Confusion matrix : 
 rug_pred    0   1
y_test           
0         100  10
1          24  45

Accuracy  = 0.8101
F1 score  = 0.7258


### Questions

**Q 5.2.1 – Evaluate the model performance.**

The model performs quite good in terms of accuracy, with a score that's higher than that of logistic regression and the CART decision tree. The F1-score has a value of 72.58%, which is pretty good and exceeds those of the other models fitted so far.

**Q 5.2.2 – Explain the meaning of `rule_cost = Length()`.**

The objective function of RUG, which we want to minimize, includes cost coefficients for the rules. If `rule_cost = Length()`, then these coefficients will be equal to the length of each rule (i.e. the conditions in the rule). If we use the length of each rule as the cost, then shorter rules will be preferred over longer ones, so the average rule length is likely to decrease. However, we may end up with a bit more rules and/or less accuracy.

## model interpretation

---
### GLOBAL

In [8]:
exp = Explainer(rug)
summary = exp.summarize_rule_metrics()

Total number of rules: 30
Average rule length: 2.00


In [9]:
rules = exp.retrieve_rule_details(list(feature_names))

RULE 0:
2.50      < Age       <= inf       or null
-inf      < Pclass_2  <= 0.50      or null
-inf      < Sex_1     <= 0.50      and not null
Class: 1
Scaled rule weight: 1.0000

RULE 1:
77.00     < Age       <= inf       and not null
0.50      < Embarked_2 <= inf       or null
Class: 1
Scaled rule weight: 1.0000

RULE 2:
13.50     < Age       <= inf       or null
-inf      < Fare      <= 26.14     or null
0.50      < Sex_1     <= inf       or null
Class: 0
Scaled rule weight: 0.8333

RULE 3:
-inf      < Age       <= 11.50     and not null
-inf      < Embarked_2 <= 0.50      and not null
Class: 0
Scaled rule weight: 0.8333

RULE 4:
-inf      < Age       <= 1.50      and not null
Class: 1
Scaled rule weight: 0.8333

RULE 5:
379.93    < Fare      <= inf       and not null
-inf      < SibSp     <= 4.50      or null
Class: 1
Scaled rule weight: 0.8333

RULE 6:
0.96      < Age       <= 2.50      or null
-inf      < Pclass_1  <= 0.50      or null
Class: 0
Scaled rule weight: 0.8333

RULE 7:


### Questions

**Q 5.2.2 – Evaluate the global interpretability of the model.**

The model generated 30 rules, which have an average rule length of 2.0. The rules come with weights that range from 1.00 to 0.08. With 30 rules, this model is less straightforward to grasp than the decision trees from earlier, however we gain a bit in classification performance.

---
### LOCAL

In [10]:
rule_coverage_metrics = exp.evaluate_rule_coverage_metrics(X_test, info=True)

Number of instances not covered by any rule: 0
Average number of rules per sample: 2.40
Average length of rules per sample: 1.88


In [11]:
i = 9
print(f'True and predicted values for sample at index {i}:')
print(y_results.loc[i,:], '\n')
print(f'Sample {i} features:')
X_test.loc[[i]]

True and predicted values for sample at index 9:
y_test      1
rug_pred    0
Name: 9, dtype: int64 

Sample 9 features:


Age  SibSp  Parch  Fare  Sex_1  Embarked_1  Embarked_2  Pclass_1  Pclass_2
9  19.0      0      0  8.05      1           0           1         0         1

In [12]:
exp.find_applicable_rules_for_samples(X_test.iloc[[i]], feature_names=list(feature_names), info=True)

Rules for instance 0
RULE 2:
13.50     < Age       <= inf       or null
-inf      < Fare      <= 26.14     or null
0.50      < Sex_1     <= inf       or null
Class: 0
Scaled rule weight: 0.8333

RULE 14:
-inf      < Fare      <= 27.14     or null
Class: 1
Scaled rule weight: 0.4167



[[2, 14]]

### Questions

**Q 5.2.3 – Take a look at the local explanation for the sample with index 9 of the test set. Explain how the prediction is made.**

Each of the rules assigns the sample to either class 0 or 1, and we use weighted majority voting among rules to make the prediction. In other words, we sum the weights of the rules for each class and the prediction is equal to the class with the higher weight. Here, Rule 2 assigns the sample to class 0, with a weight of $0.83$. The other rule (Rule 14) assigns the sample to class 1, with a weight of 0.42. As the weight for class 0 is higher than the weight of class 1, the prediction for the sample is class 1.

**Q 5.2.4 – Explain the main difference between RUG and the algorithm here that produces a decision list.**

RUG returns a decision set, and for prediction we use all of the rules that apply to the sample. The individual predictions are weighted using the rule weights, and the class with the higher summed weight will be assigned to the sample. 

In a decsision list like here, a sample is classified with only one rule, namely the first in the list where the sample satisfies the conditions.